### DISCUSSION TITLE
Why isn't 'include_background' named 'include_first_channel'?

### DISCUSSION THREAD

**Discussion by ellisdg** [Project-MONAI/MONAI#6915] [Q&A]
In many of the loss classes and metrics in MONAI, the creators have put in an argument named `include_background`. To my understanding, what this argument does is include the first channel in the computations, if True, and exclude the first channel in the computations, if False.

I have a few issues with the naming of the `include_background` argument:
1) The name is ambiguous and does not reflect what the argument actually does.
2) The background channel could be any of the channels, and does not necessarily have to be the first channel.
3) When there is no background channel, you should specify `include_background=True`, which is counter-intuitive because there is no background to be included.

Therefore, why isn't the argument called `include_first_channel` instead? This name says what the `include_background` argument actually does.

**Comment by KumoLiu** [comment]
Hi @ellisdg, thanks for your interest and suggestion here.
We used 'include_background' to specify whether we calculated with or without the first category, which is usually assumed to be background. Using 'include_first_channel' would be misleading because what we want to do is determine whether or not to include background. Does it make sense?
 https://github.com/Project-MONAI/MONAI/blob/a4e4894dca25f5e87b9306abfc472805f92b69da/monai/metrics/meandice.py#L30-L32

Additionally, when only single-channel class indices exist, we can also specify whether to include background.
https://github.com/Project-MONAI/MONAI/blob/a4e4894dca25f5e87b9306abfc472805f92b69da/monai/metrics/meandice.py#L253-L258

I hope it helps, thanks!

**Reply by ebrahimebrahim** [reply]
As a MONAI user when I saw `include_background`, I had to check at the documentation to make sure that by "background" it meant the zeroth channel. `include_first_channel` or `include_zeroth_channel` would have been clearer to me.

**Reply by ellisdg** [reply]
@KumoLiu, thank you for your reply!

> As a MONAI user when I saw `include_background`, I had to check at the documentation to make sure that by "background" it meant the zeroth channel. `include_first_channel` or `include_zeroth_channel` would have been clearer to me.

I agree with @ebrahimebrahim. Either of those suggestions would increase the clarity and ultimately reduce the amount of errors that users make. `include_channel_zero` could be another option.

> Using 'include_first_channel' would be misleading because what we want to do is determine whether or not to include background.

The `include_background` flag is misleading when you have a single or multi-channel output and there is no background channel, because setting `include_background=True` does not actually include the background.

> Additionally, when only single-channel class indices exist, we can also specify whether to include background.

The example code you provided reinforces my point that when we say `include_background` we are talking about the first channel (or `zeroth_channel` as @ebrahimebrahim suggested). When `include_background` is `True`, then `first_ch` is set to 0, which forces the first channel to be included. When `include_background` is `False`, then `first_ch` is set to 1, which forces the first channel to be excluded. I think getting rid of the `include_background` flag altogether and letting the user specify the first channel would be more intuitive than the `include_background` flag. It doesn't matter if we are talking about actual channels or label indices, the code you provided uses the `first_ch` variable for both. 

In my opinion, the `first_ch` variable would be better named something along the lines of `first_evaluated_channel`, `first_eval_ch`, or `start_channel`, because the first channel doesn't ever change from being channel zero, but the first channel being evaluated can change.

**Reply by KumoLiu** [reply]
Thanks, @ellisdg and @ebrahimebrahim for your suggestions.
If that is the general confusion among MONAI  users, I believe we should reconsider the name. Let's bring it up to more people here. cc @wyli @ericspod @Nic-Ma @vikashg

**Comment by ericspod** [comment]
It has been convention that background was always channel 0 and this has made sense for the segmentation that we've been doing. I can see other applications where there is no background as described here so changing the name probably should be done, eg. "include_channel_zero" or something like that. This variable name is one of those, shall we say, "historical" parts of MONAI. I would still be explicit in documentation and elsewhere that the background channel is channel 0 if any.

**Reply by ellisdg** [reply]
This is might be a naive question, but what is the benefit to having a background channel if it isn't included in the loss function?

**Reply by ericspod** [reply]
If you have a multi-class segmentation problem where the segments are relatively small sometimes the loss function won't penalise the network enough if it predicts all background. The background signal in this case tends to dominate and the network can forget about smaller classes. You still need a background channel of course because every pixel/voxel must have a prediction.  For example, cardiac MR segmentation of the ventricles is where I've encountered this behaviour.

**Reply by ellisdg** [reply]
Thank you for your response!

> You still need a background channel of course because every pixel/voxel must have a prediction.

You can also use a sigmoid activation function and then assign the background to any voxels that don't pass the activation threshold. That is the strategy taken in this MONAI tutorial for BRATS tumor segmentation:
https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/brats_segmentation_3d.ipynb

In this case the label channels are tumor core, whole tumor, and enhancing tumor, and there is no background channel. So I guess I am still confused as to why the default behavior should be that all models have a background channel.

**Reply by ericspod** [reply]
Using sigmoid is something we've done with binary segmentation where the prediction tensor has only 1 channel. For doing this with multi-class I guess it's possible but not what I've seen done and may not perform well with Dice or other losses.

**Comment by ChenglongWang** [comment]
I believe that using "include_channel_zero" as an alternative to "include_background" may not be the best choice due to the loss of argument semantics. In contrast, the name "include_background" directly conveys the purpose of this argument to the user. The core issue of original post is that the channel of background is confusing. One potential solution could be to introduce an optional argument, such as `background_channel: int | Sequence[int] = 0`, to clarify this aspect.  BTW, argument such as `include_channel: int | Sequence[int] = 0` could be better than "include_channel_zero" .

**Reply by ericspod** [reply]
Thanks for the feedback, we'll keep this in mind for future revisions.
